### ISSUE TITLE
`install.md` Review

### ISSUE THREAD

**Issue by gdvallance** [spiritumduo/spiritumDuo#13]
### Overall

- **FAILED** to get a working instance using the instructions. Could not get the database to migrate (see below).
- Pick issue up at next sprint meeting ...

### Installation Pre-requisites
- Clear instructions
- Tested on an Ubuntu 20.04 VM with *only* and editor, `git`, `docker` & `docker-compose` installed

### Configuring Environment Variables
- A bit tedious duplicating *but* the instructions were clear
- Missing an instruction to create the `nginx` `.env` file ... Eg. 6. `nginx/.env.example` at the top of the section. The description *is* covered later.

### Building `node` modules
- I had an initial failure and then realised I was on a `vpn` and I had an issue resolving [https://registry.yarnpkg.com/](https://registry.yarnpkg.com). It manifested with this error:

```bash
Step 4/5 : RUN yarn install --development --silent
 ---> Running in e87d9f47b409
warning Pattern ["@apollo/client@latest"] is trying to unpack in the same destination "/usr/local/share/.cache/yarn/v6/npm-@apollo-client-3.5.5-ce331403ee5f099e595430890f9b510c8435c932-integrity/node_modules/@apollo/client" as pattern ["@apollo/client@>=2.0.0"]. This could result in non-deterministic behavior, skipping.
error An unexpected error occurred: "https://registry.yarnpkg.com/babel-loader/-/babel-loader-8.1.0.tgz: getaddrinfo ENOTFOUND registry.yarnpkg.com".
The command '/bin/sh -c yarn install --development --silent' returned a non-zero code: 1
Unable to find image 'sd-frontend:latest' locally
docker: Error response from daemon: pull access denied for sd-frontend, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.

``` 

- Once off `vpn` it ran fine **BUT** a log of warnings, which *might* be concerning. [If you could fix the errors good, but I suspect resolving the dependencies is nigh on impossible ...

### Configure `docker-compose.dev.yml`
- A bit confusing in that I missed the instruction to create a `docker-compose.dev.yml` file from the `docker-compose-dev.yml.example` file. See below. (Or I was confused. Could be both. I note it because I experienced it ...)

### Build all containers
- The instruction to move to the root directory might be clear most, but there are some where it might not be. Could be worth earlier in the document defining terms such as what is meant by "root directory" etc.
- Running `docker-compose -f docker-compose.dev.yml up -d --build` created an error. To address need to say at the beginning to copy and rename `docker-compose.dev.yml.example` to `docker-compose.dev.yml`. E.g. 

1. Move to the project's root directory
2. Create a dev `docker-compose` file by running: `cp docker-compose.dev.yml.example docker-compose.dev.yml`
3. Run `docker-compose -f docker-compose.dev.yml up -d --build

- **NB** Reviewing the instructions again I see that this step was covered in Configure `docker-compose.dev.yml` section. So, for whatever reason I missed it which could mean a bunch of things -- like it was late and I am tired and/or the documentation could be clearer.

- Had an initial failure because there had been no instruction to create a `.env` file in the `nginx` section.

### Import database schemas

- Big chunks of text are better in a separate code-block because it is all-to-easy to miss copy. I did, although I was copying from the source rather than from the rendered. Easier of course from rendered. Still thing the point applies though.
- Copying from rendered still go an error about the "database "spiritum_duo" [which I called it in the .env file] does not exist". Looking at it again, it confirms the point I made that the instructions would be better in code blocks. I missed the and in the two instructions. E.g.

```bash
docker exec -ti sd-backend bash -c "chmod +x ./bin/container-migrate-alembic && ./bin/container-migrate-alembic"

```
and

```bash
docker exec -ti sd-pseudotie bash -c "chmod +x ./bin/container-migrate-alembic && ./bin/container-migrate-alembic"

```

- Still get an error that database does not exist. I note that there is an instruction that the containers need to be running before the database can be migrated, *without* instructions as to get them running. I believe that this would be a good idea, or some instruction as to check that they *are*. 
- Furthermore, the containers *are* running so not sure what is going on, and there is a reboot-loop for the `sd-nginx-dev` container, so there is an issue there. i.e. it keeps restarting.

- Error reproduced below:

```bash
docker exec -ti sd-backend bash -c "chmod +x ./bin/container-migrate-alembic && ./bin/container-migrate-alembic"
/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/base.py:299: SAWarning: Can't validate argument 'native_enum'; can't locate any SQLAlchemy dialect named 'native'
  util.warn(
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2336, in _wrap_pool_connect
    return fn()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 304, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 778, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 495, in checkout
    rec = pool._do_get()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 241, in _do_get
    return self._create_connection()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 309, in _create_connection
    return _ConnectionRecord(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 440, in __init__
    self.__connect(first_connect_check=True)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 661, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 656, in __connect
    connection = pool._invoke_creator(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/strategies.py", line 114, in connect
    return dialect.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 508, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.9/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: FATAL:  database "spiritum_duo" does not exist


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/alembic/config.py", line 588, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/usr/local/lib/python3.9/site-packages/alembic/config.py", line 582, in main
    self.run_cmd(cfg, options)
  File "/usr/local/lib/python3.9/site-packages/alembic/config.py", line 559, in run_cmd
    fn(
  File "/usr/local/lib/python3.9/site-packages/alembic/command.py", line 227, in revision
    script_directory.run_env()
  File "/usr/local/lib/python3.9/site-packages/alembic/script/base.py", line 563, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/usr/local/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 92, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/local/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 108, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "alembic/env.py", line 89, in <module>
    run_migrations_online()
  File "alembic/env.py", line 77, in run_migrations_online
    with connectable.connect() as connection:
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2263, in connect
    return self._connection_cls(self, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 104, in __init__
    else engine.raw_connection()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2369, in raw_connection
    return self._wrap_pool_connect(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2339, in _wrap_pool_connect
    Connection._handle_dbapi_exception_noconnection(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1583, in _handle_dbapi_exception_noconnection
    util.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2336, in _wrap_pool_connect
    return fn()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 304, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 778, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 495, in checkout
    rec = pool._do_get()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 241, in _do_get
    return self._create_connection()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 309, in _create_connection
    return _ConnectionRecord(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 440, in __init__
    self.__connect(first_connect_check=True)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 661, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 656, in __connect
    connection = pool._invoke_creator(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/strategies.py", line 114, in connect
    return dialect.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 508, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.9/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) FATAL:  database "spiritum_duo" does not exist

(Background on this error at: http://sqlalche.me/e/13/e3q8)
/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/base.py:299: SAWarning: Can't validate argument 'native_enum'; can't locate any SQLAlchemy dialect named 'native'
  util.warn(
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2336, in _wrap_pool_connect
    return fn()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 304, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 778, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 495, in checkout
    rec = pool._do_get()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 241, in _do_get
    return self._create_connection()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 309, in _create_connection
    return _ConnectionRecord(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 440, in __init__
    self.__connect(first_connect_check=True)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 661, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 656, in __connect
    connection = pool._invoke_creator(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/strategies.py", line 114, in connect
    return dialect.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 508, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.9/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: FATAL:  database "spiritum_duo" does not exist


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/alembic/config.py", line 588, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/usr/local/lib/python3.9/site-packages/alembic/config.py", line 582, in main
    self.run_cmd(cfg, options)
  File "/usr/local/lib/python3.9/site-packages/alembic/config.py", line 559, in run_cmd
    fn(
  File "/usr/local/lib/python3.9/site-packages/alembic/command.py", line 320, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.9/site-packages/alembic/script/base.py", line 563, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/usr/local/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 92, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/local/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 108, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "alembic/env.py", line 89, in <module>
    run_migrations_online()
  File "alembic/env.py", line 77, in run_migrations_online
    with connectable.connect() as connection:
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2263, in connect
    return self._connection_cls(self, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 104, in __init__
    else engine.raw_connection()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2369, in raw_connection
    return self._wrap_pool_connect(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2339, in _wrap_pool_connect
    Connection._handle_dbapi_exception_noconnection(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1583, in _handle_dbapi_exception_noconnection
    util.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2336, in _wrap_pool_connect
    return fn()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 304, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 778, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 495, in checkout
    rec = pool._do_get()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 241, in _do_get
    return self._create_connection()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 309, in _create_connection
    return _ConnectionRecord(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 440, in __init__
    self.__connect(first_connect_check=True)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 661, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 656, in __connect
    connection = pool._invoke_creator(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/strategies.py", line 114, in connect
    return dialect.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 508, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/usr/local/lib/python3.9/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) FATAL:  database "spiritum_duo" does not exist

(Background on this error at: http://sqlalche.me/e/13/e3q8)

```

#### `install.md`
- The formatting of `install.md` does not pass `markdownlint` on `VSCode`. **LOTS** of errors.

**Comment by jcf54** [comment]
Thank you for reviewing this, incredibly useful feedback. I've just committed changes to resolve the issues you've listed above. 

Namely,
- Added an instruction to create the nginx env file in the section `Configuring environmental variables`
- Added a table with the definition of `project's root directory` or wtte in the context used in this document
- Placed the two commands for database migration in their own code blocks to avoid potential confusion having them so close to each other
- Delinted the document
- Resolved an issue re nginx wherein it will enter a reboot loop and constantly error on Linux hosts
- Changed the default envvars in postgres' env file and backend's env file to make the database name suggestion consistent 

I appreciate there are a couple of areas we can improve programmatically (for example, running the migrations and then having to restart the containers could be avoided) and I've marked them down on our Jira board. 

I'm also interested in making a Python script to setup the containers (and do the mundane stuff like copying env files) that will complement this documentation. I've added that to our Jira board too.

Please do let me know if I've missed anything or if there's further room for improvement

**Comment by gdvallance** [comment]
## Further comments

- Still failed but much closer.

### Configuring Environment Variables > backend/.env

- The directions do not make sense in this section:

```markdown
- DATABASE_USERNAME
    - This should be equal to POSTGRES_USER in backend\.env

- DATABASE_PASSWORD
    - This should be equal to POSTGRES_PASSWORD in backend\.env

```

- The comments should be that they are equal to the POSTGRES_USER & POSTGRES_PASSWORD in **`postgres\.env`**

### File `pseudotie\.env.example`

- The configuration is incorrect. There are defaults are preset as `postgres` for DATABASE_USERNAME & DATABASE_PASSWORD whereas they should be set to the values in `backend\.env`. The variable used in `postgres\.env.example` `backend_db` could be adapted and say here `backend_database_username` and `backend_database_password` respectively.
- If you don't you cannot get the migrations to work.

### Import Database Schemas

- The command `docker-compose down` fails because there are docker containers still up that mean the network cannot cannot be removed. For the command to work properly we need to stop `sd-frontend` and `sd-fontend-sb` manually -- or at least I did
- It leaves `sd-frontend` and `sd-frontend-sb` containers up 
- If you stop them to successfully run the command you then need to remember to restart them before the next command `docker-compose -f docker-compose-dev.yml up -d` is executed but *that* fails. It seems better to ignore the errors ...

### Rebooting `sd-nginx-dev`

- The docker container remains in continual restart.

### `docker-compose -f docker-compose.dev.yml up -d`

- I got the following ... ? successful ...

```bash
Testing connection...
Connection successful!
User inserted
Username: user0
Password: QS7k34j1SOde
^CTraceback (most recent call last):
  File "/app/src/manage.py", line 249, in <module>
    loop.run_until_complete(insert_test_data())
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 1854, in _run_once
    event_list = self._selector.select(timeout)
  File "/usr/local/lib/python3.9/selectors.py", line 469, in select
    fd_event_list = self._selector.poll(timeout, max_ev)
KeyboardInterrupt

```

### Interesting Warnings

- In migration and starting up the containers for the first time you get:

```bash
/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/base.py:299: SAWarning: Can't validate argument 'native_enum'; can't locate any SQLAlchemy dialect named 'native'
  util.warn(

```

### General comments

- Doesn't say where the application will  be available etc. on successful installation

**Comment by jcf54** [comment]
Thanks for doing another number on this. I'll sort the wording for the envvar definitions shortly + add a note where the webapp's listening

I'm not sure why the frontend containers are getting orphaned, I hope to get a good look at it by Friday.
Regarding the nginx container restarting, did you happen to take note of the error the container throws when it crashes?

**Comment by gdvallance** [comment]
I have fired up the VM which I have deployed the app on and interestingly the following containers did **not** restart:
1. `sd-frontend`
2. `sd-frontend-sb`
3. `sd-backend`
4. `sd-speudotie`

I have not looked into why -- don't have the capacity presently.

The following start up fine:
1. `wordpress:5.8.2-fpm-alpine`
2. `mysql:8.0
3. `postgres:14.0-alpine`

And finally I get a reboot loop for:
1. `sd-nginx-dev`

The output of `docker logs sd-nginx-dev was repetitions of --

```bash
2022/03/02 10:31:56 [emerg] 1#1: host not found in upstream "host.docker.internal" in /etc/nginx/conf.d/default.conf:25
nginx: [emerg] host not found in upstream "host.docker.internal" in /etc/nginx/conf.d/default.conf:25

```

**Comment by jcf54** [comment]
I've just corrected the issue re envvar definitions and the confusion around what should be what. Also updated the default envvars to be more comprehensive re database connection info.

I made changes to the nginx configuration in the docker-compose file a couple of days ago that should have corrected the `host.docker.internal` error so not sure why that's still being an issue.

Do you mind sending me the versions of these?
- Linux
- Docker
- Docker compose

When running either `manage,py` or `manage-demo.py`, the script will generate user credentials then go silent until the operations are complete. I'll make it more obvious when it's doing this.

**Comment by gdvallance** [comment]
## Deployment VM details

- **OS:** Ubuntu 20.04.4 LTS
- **Docker:** Docker version 20.10.12 build e91ed57
- **Docker Compose:** version 1.29.2 build 5becea4c

**Comment by jcf54** [comment]
Re nginx errors, would you mind confirming that your `docker-compose.dev.yml` file's NGINX service has this attribute?

```
extra_hosts:
      - "host.docker.internal:host-gateway"
```

**Comment by gdvallance** [comment]
Progress.

- To answer the question, the `host.docker.internal:host-gateway` line was NOT in the `docker-composel.dev.yml`. It now is ...

- From a clean slate (after deleting everything), *almost* have working example ... The error I get now on `http://localhost` is "Error establishing a database connection". This is despite running the test data insertion and getting success *and* restarting everything again. 
- `https://localhost` does not work.

- In running the database migrations and during `docker-compose -f docker-compose.dev.yml up -d` we get: 

`/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/base.py:299: SAWarning: Can't validate argument 'native_enum'; can't locate any SQLAlchemy dialect named 'native'
  util.warn(` 

and

`/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/base.py:299: SAWarning: Can't validate argument 'native_enum'; can't locate any SQLAlchemy dialect named 'native'
  util.warn(` respectively.

- When you execute `docker-compose down` you get the following. Hard to know if important ... 

`WARNING: Found orphan containers (sd-frontend-sb, sd-frontend) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up.` 

and

 `WARNING: Found orphan containers (sd-frontend-sb, sd-frontend) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up.`

**Comment by jcf54** [comment]
That's great! It's working! 

Our stack comes with Wordpress (we should probably remove that, it's for the landing page on https://spiritumduo.com). I forgot to put a note in the instructions about this - I will shortly, I'm just fixing some UI stuff at the mo.

The app will be listening on http://localhost/app

**Comment by jcf54** [comment]
I've just changed the configurations of the containers so they will restart unless they have been stopped deliberately. This should resolve your first point. I'll get it pushed and the docs updated tonight. 

I don't think I've seen nginx throw 504 recently? I think that happens when the frontend container hasn't finished building, or has errored. 
If you check the logs of `sd-frontend` do you mind checking if it's successfully built? I think the last output from it should be some stuff from the linter

**Comment by gdvallance** [comment]
FFS.

I just tried it *now* after many repeats and it works ... 

I litterally just refreshed it once more time after messaing above and it worked <sigh>.

**Comment by gdvallance** [comment]
And while I am thinking about it, if it is the case it takes a while for the containers to settle down etc. and work we need to warn testers that this could take a while OR/ create something that will indicate after you run the scripts that everything is now up and ready along with how to access it. Something to discuss perhaps.

**Comment by jcf54** [comment]
No worries. I'll write a note about the delays in the docs. I'd be lying if I said it hadn't caught me out once or twice too. I wanted to look at making a script to automate the installation process, it's on the backlog at the moment. That could be handy.

I thought we fixed the issue with pagination, we'll have to see if it's dropped through the cracks. I've added it to our backlog so we'll look into it soon.

As a side note, you can run the manage scripts as many times as you like, it nukes all the data in the database before re-adding fresh data.

**Comment by jcf54** [comment]
Alrighty, I've pushed a few things to the documents/what we've discussed.

1. I've fixed the issue that generated this error `/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/base.py:299: SAWarning: Can't validate argument 'native_enum'; can't locate any SQLAlchemy dialect named 'native'`
2. Corrected envvar definitions for the database parameters in both `backend/.env` and `pseudotie/.env`
3. Added note where the app listens
4. Set containers to restart unless stopped (either manually or in event of a crash), added note to support this
5. Added a note that frontend containers may take a couple of minutes to listen/for someone to connect

I also think in changing the container restart policy, we've fixed the orphaned-container problem too. I could be wrong but I can't reproduce the error now

**Comment by gdvallance** [comment]
Excellent. I have just run through the instructions again from scratch -- apart from the images, and it a much nicer experience with only one distracting message in the `docker-compose down` in the "Import Database Schemas" section. Specifically you get:

```bash
ERROR: error while removing network: network spiritumduo_app-network id 6708637be753311a9ffe22dc1f28bcce90f8d23f6e14aae7dc9ec2b42e729efc has active endpoints

```

I know that this does not mean much but if folks have made a mis-step earlier or don't know much about docker they could waste time trying to troubleshoot. I would add a note to the effect that if you get this ERROR then please ignore.

I would also add a comment in the "Insert Test Data Section" saying to use the generated credentials to log in to the system once it comes up and that if necessary you can run this script as many times as necessary to get new login credentials.

Pagination may or may not be working correctly. Is it the case you have two identical pages? If so, you *really* want to have 2 distinct pages if you have 2 pages. Otherwise, only insert say 1.5 pages worth of data ...

We are basically there :-D.

**Comment by jcf54** [comment]
I think containers are still getting orphaned which raises that network error. I've been looking at it this morning and I can't find anything concrete. I'd rather fix it entirely over adding a couple of workarounds so will keep looking.
I've added a note regarding re-running the test data script, will push later. I'm still not sure what's going on with pagination, before iirc it was a caching problem with the GraphQL client library? Will look either tonight or Friday if I can find some time

**Comment by jcf54** [comment]
I've narrowed down the issue re orphaned containers and the network issue you described. `docker-compose.yml` is the default compose file, and we're using the file `docker-compose.dev.yml` for the dev compose configuration. We specify this on compose up `docker-compose -f docker-compose.dev.yml up --build`, but not on compose down. I've just corrected this and added a note for running the test script multiple times
